src/PhyloData.cpp

```diff
@@ -0,0 +1,617 @@
+  }
+
+  // Store the new marginal probability matrix in the Smooshable object.
+  sp_cast->AuxUpdateMarginal(new_marginal);
+};
+
+
+// Pile Functions
+
+
+/// @brief Marks all the Smooshishs in `vdj_pile_` as dirty.
+void PhyloData::MarkPileAsDirty() const {
+  // Iterate over the SmooshishPtr's and mark each as dirty.
+  for (std::size_t i = 0; i < vdj_pile_.size(); i++) {
+    vdj_pile_[i]->MarkAsDirty();
+  }
+};
+
+
+/// @brief Cleans the dirty Smooshishs in `vdj_pile_` by recomputing marginal
+/// probability matrices for raw Smooshables according to the current phylogeny.
+///
+/// Note that this function recomputes marginal probability matrices for dirty
+/// Smooshables only (and not for dirty Chains).  The probability matrices in
+/// Chain objects are computed from the corresponding matrices in raw Smooshable
+/// objects.  Thus, once the dirty Smooshables have been updated, the downstream
+/// dirty Chains can be updated too (although this does not happen in this
+/// function).
+///
+/// Ideally, we would recurse over each SmooshishPtr in `vdj_pile_` and update
+/// marginal probability matrices during the recursion.  This is infeasible in
+/// our setting because most of the information needed to recompute marginal
+/// probability matrices is stored in the `PhyloData` class; trying to pass such
+/// a class to the updating function would introduce a circular dependency.
+/// Thus, we recurse over each SmooshishPtr in `vdj_pile_` and extract out the
+/// dirty Smooshables in the given Smooshish.  These dirty Smooshables are then
+/// updated in `PhyloData` and all Smooshishs in the given Smooshish we recursed
+/// over are considered clean (remember that Chains being clean only depend on
+/// the upstream Smooshables being clean).
+void PhyloData::CleanPile() const {
+  // Iterate over the SmooshishPtr's and clean each one.
+  for (std::size_t i = 0; i < vdj_pile_.size(); i++) {
+    // Find the dirty Smooshable(s) for the given Smooshish.
+    std::vector<SmooshishPtr> dirty_smooshables =
+        FindDirtySmooshables(vdj_pile_[i]);
+
+    // Update the marginal probability matrices in all the dirty Smooshable(s).
+    for (std::size_t j = 0; j < dirty_smooshables.size(); j++) {
+      UpdateMarginal(dirty_smooshables[j]);
+    }
+
+    // Mark all the Smooshishs contained within the given Smooshish as clean.
+    vdj_pile_[i]->MarkAsClean();
+  }
+};
+
+
+// Auxiliary Functions
+
+
+/// @brief Caches the xMSA index vector for a given germline gene.
+/// @param[in] germ_ptr
+/// A pointer to an object of class Germline.
+/// @param[in] left_flexbounds_name
+/// The name of the left flexbounds, which is a 2-tuple of MSA positions
+/// providing the bounds of the germline's left flex region.
+/// @param[in,out] xmsa_ids
+/// A map identifying already cached xMSA site indices.
+void PhyloData::CacheGermlineXmsaIndices(
+    GermlinePtr germ_ptr, std::string left_flexbounds_name,
```

Comment:
For some detail between MSA-xMSA conversion, see https://github.com/matsengrp/linearham/issues/39#issuecomment-272086562.